src/host/_stream.cpp

```diff
@@ -703,20 +631,24 @@ NT_CATCH_RETURN()
 [[nodiscard]] HRESULT ApiRoutines::WriteConsoleWImpl(IConsoleOutputObject& context,
                                                      const std::wstring_view buffer,
                                                      size_t& read,
-                                                     std::unique_ptr<IWaitRoutine>& waiter) noexcept
+                                                     CONSOLE_API_MSG* pWaitReplyMessage) noexcept
 {
     try
     {
         LockConsole();
         auto unlock = wil::scope_exit([&] { UnlockConsole(); });
 
-        std::unique_ptr<WriteData> writeDataWaiter;
-        RETURN_IF_FAILED(WriteConsoleWImplHelper(context.GetActiveBuffer(), buffer, read, writeDataWaiter));
-
-        // Transfer specific waiter pointer into the generic interface wrapper.
-        waiter.reset(writeDataWaiter.release());
+        auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
+        if (WI_IsAnyFlagSet(gci.Flags, (CONSOLE_SUSPENDED | CONSOLE_SELECTING | CONSOLE_SCROLLBAR_TRACKING)))
+        {
+            std::ignore = ConsoleWaitQueue::s_CreateWait(pWaitReplyMessage, new WriteData(context, std::wstring{ buffer }, gci.OutputCP));

```

Comment:
raw use of `new`: this and the others are only provably safe if `s_CreateWait` is noexcept; otherwise we _could_ construct a waiter and it would be leaked if it throws, right?

Reply:
This is correct, but also already the case, as the previous code used `release()` to hand off a raw unowned pointer to `s_CreateWait`. Effectively, the safety of the function has not changed: It's still incorrect. This is also why I'm ignoring the returned `HRESULT`, because it's similarly unuseful.

I would prefer refactoring this as part of "Server v2" aka In-Process-ConpTY.